fastcrypto/src/ristretto255.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implementations of the [ristretto255 group](https://ristretto.group/) which is a group of
+//! prime order 2^{252} + 27742317777372353535851937790883648493.
+
+use std::ops;
+
+use curve25519_dalek_ng;
+use curve25519_dalek_ng::traits::Identity;
+use once_cell::sync::OnceCell;
+use serde::{de, Deserialize, Serialize};
+
+use crate::hash::HashFunction;
+use crate::{error::FastCryptoError, traits::ToFromBytes};
+
+/// Represents a scalar modulo the order of the ristretto group which is 2^{252} + 27742317777372353535851937790883648493.
+pub struct RistrettoScalar(curve25519_dalek_ng::scalar::Scalar);
+
+impl RistrettoScalar {
+    /// Attempt to create a new scalar from the given bytes in canonical representation.
+    pub fn from_canonical_bytes(bytes: [u8; 32]) -> Option<RistrettoScalar> {
+        curve25519_dalek_ng::scalar::Scalar::from_canonical_bytes(bytes).map(RistrettoScalar)
+    }
+
+    /// Create a scalar from the given `u64`.
+    pub fn from(value: u64) -> RistrettoScalar {
+        RistrettoScalar(curve25519_dalek_ng::scalar::Scalar::from(value))
+    }
+
+    /// Create a scalar from the low 255 bits of the given 256-bit integer.
+    pub fn from_bits(value: [u8; 32]) -> RistrettoScalar {
+        RistrettoScalar(curve25519_dalek_ng::scalar::Scalar::from_bits(value))
+    }
+}
+
+/// Represents a point in the Ristretto group for Curve25519.
+#[derive(Debug)]
+pub struct RistrettoPoint {
+    point: curve25519_dalek_ng::ristretto::RistrettoPoint,
+    bytes: OnceCell<[u8; 32]>,
+}
+
+impl RistrettoPoint {
```

Comment:
Do we need scalar to point multiplication + is_identity as well for a complete api?

Reply:
Scalar to point multiplication and map2point is already there as `impl ops::Mul<RistrettoPoint> for RistrettoScalar` and `pub fn from_uniform_bytes(bytes: &[u8; 64]) -> Self` resp. Or is that not what you are referring to? And is_identity should definitely be added.

Reply:
Indeed re scalar to point.
Map to point is a special function, referred to hash to point sometimes and we usually use the Elligator2 hash function.
Need to check if the Ristretto lib implements Elligator or Elligator2 in https://docs.rs/curve25519-dalek/latest/curve25519_dalek/ristretto/index.html#random-points-and-hashing-to-ristretto

Reply:
The `from_uniform_bytes` method in the ristretto lib uses the Elligator map so I think it's already there (see the `hash_from_bytes` function), but maybe it should be renamed. I don't know if it's Elligator or Elligator2 though. It's referred to in the documentation as the "Elligator map" so I'm guessing it's the former, but let me just check.

Reply:
![image](https://user-images.githubusercontent.com/25659162/202135653-2496279f-c588-43e9-9ac3-29dcb687406f.jpeg)